Comment Content Block: Add typography, color and padding support

DAreRodz opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
This PR continues the implementation of the Post Comment Content block that already existed. It adds support for text alignment, font size, family, style and weight, line height, colors (including links and gradients), and vertical/horizontal padding.

Will close https://github.com/WordPress/gutenberg/issues/30574.

## How has this been tested?

1. Add a post comment block (using an existing comment ID).
2. Add an inner post comment content block.
3. Confirm that the color and typography options work.
4. Confirm that the text align option works.
5. Confirm that the padding option works.
6. Create a template containing just the Comment Content block, and confirm that the fallback message (`The content of the comment.`) is shown.

## Types of changes

Enhancement

## Checklist:
- [x] My code is (manually) tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

DAreRodz committed:
Comment Content Block: add more `supports` settings

Settings are:
* Typography and color settings (same as the Paragraph block)
* Layout settings (same as Post Content block)

mtias commented:
Commenting here but we should transfer to the general tracking issue — the "Post" prefix is a bit pointless, we should simplify the naming of all these blocks to be "Comment Content", "Comment Author" etc.

ntsekouras commented on packages/block-library/src/post-comment-content/block.json:
```diff
@@ -8,6 +8,14 @@
+			"lineHeight": true
+		},
+		"html": false,
+		"__experimentalLayout": true
```
I think we will not need to add the `__experimentalLayout` for this block. Layout is for container blocks that enables controlling `width/height`.

DAreRodz replied:
Thanks, @ntsekouras. I did that because @SantosGuillamot suggested adding that specific setting. I'll talk with him to clarify that.

SantosGuillamot replied:
I suggested adding that because it's something supported in the [Post Content block](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/post-content/block.json), and I thought it could be a good idea to follow a similar pattern. It seemed to me that it would add more flexibility for editing the layout, but I don't have a strong opinion here, so no problem if you feel it's better to change it 🙂

mtias replied:
@scruffian @MaggieCabrera @jffng curious what has come up in themes as uses here.

scruffian replied:
I've not seen a real life use case where we'd want to use this outside of a post comment block.

ntsekouras replied:
I think `layout` would make more sense in the parent blocks like `Comment Loop`

jameskoster commented:
>  I understand that this happens when the comment ID is not specified in the parent block, am I right?

Correct. Example: editing the single post template in isolation, with no post data.

> What should the block show when the comment ID is not correct? Is it OK showing the same message?

I presume you're referring to this interface in the current "Post Comment" block:

<img width="372" alt="Screenshot 2021-09-28 at 15 44 26" src="https://user-images.githubusercontent.com/846565/135110270-16d0be0a-0416-4adc-b6a5-310f7fb9ce9f.png">

I don't think we should offer this functionality in the Comment Content block since it is strictly a template-only block, IE it will always have an ID passed to it by the queried post.

> Regarding the icon, I guess I have to create a new one inside @wordpress/icons using the design, right?

Perfect! Here's the SVG:

```
<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M6.68822 16.625L5.5 17.8145L5.5 5.5L18.5 5.5L18.5 16.625L6.68822 16.625ZM7.31 18.125L19 18.125C19.5523 18.125 20 17.6773 20 17.125L20 5C20 4.44772 19.5523 4 19 4H5C4.44772 4 4 4.44772 4 5V19.5247C4 19.8173 4.16123 20.086 4.41935 20.2237C4.72711 20.3878 5.10601 20.3313 5.35252 20.0845L7.31 18.125ZM16 9.99997H8V8.49997H16V9.99997ZM8 14H13V12.5H8V14Z" fill="black"/>
</svg>
```

> I'm not sure, but these settings don't seem quite possible to be added as that would modify the comment content, if I'm not mistaken.

Good point, these should probably be omitted :)

ntsekouras commented:
>I don't think we should offer this functionality in the Comment Content block since it is strictly a template-only block, IE it will always have an ID passed to it by the queried post.

I agree with @jameskoster , but let's change this when we have the `comments loop` block. What do you think?

jameskoster commented:
~Hmm, this is a new block, no? So no harm leaving it out?~

~To be clear, we should keep that interface for the Post Comment block, but not have it for the Comment Content block.~

Edit: I understand now. Imo we should update the Post Comment block to check the ID when the user adds it, and throw an "comment not found" error (or similar) if they input a bad ID.

DAreRodz commented:
> Imo we should update the Post Comment block to check the ID when the user adds it, and throw an "comment not found" error (or similar) if they input a bad ID.

Makes completely sense, I'll do that. Thanks for all your feedback, guys. :raised_hands:

DAreRodz committed:
Comment Content: Add block icon

DAreRodz committed:
Comment Content: Show placeholder when no context

DAreRodz committed:
Comment Content: Add jsdoc to `edit` component

DAreRodz committed:
Comment Content: Remove the layout setting

ntsekouras commented on packages/icons/src/library/comment-content.js:
```diff
@@ -0,0 +1,16 @@
+const commentContent = (
+	<SVG viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
+		<Path
+			fill-rule="evenodd"
```
You have to change these props to `fillRule` and `clipRule` as `Path` is a React component. This way it produces React `Invalid DOM property` warnings.

DAreRodz replied:
You're completely right. Thanks for noticing it. :+1:

ntsekouras commented on packages/block-library/src/post-comment-content/edit.js:
```diff
@@ -15,13 +24,23 @@ export default function Edit( { context: { commentId } } ) {
 	);
+
+	// Show a placeholder when there is no context.
+	if ( ! commentId ) {
```
If no `commentId` exists the message should be something like `no comment exists`.
If `content` is `undefined` has two cases: 
1. still loading
2. comment doesn't exist (this can be handled in a separate PR though and has to do with `useEntityProp`)

The below `if ( ! content?.rendered )` check (I had hastily added yesterday) is not correct as it should be shown when the comment exists and there is no content.

DAreRodz replied:
> If no `commentId` exists the message should be something like `no comment exists`.

Maybe (or probably :sweat_smile:) I'm wrong, but from what @jameskoster wrote in [this comment](https://github.com/WordPress/gutenberg/pull/35183#issuecomment-929304277), I understood that, if you don't have `commentId` defined, it could be one of these two cases:

1. You haven't defined `commentId` in the parent block (i.e., the Comment block). In that case, I assumed that the block responsible for showing a message like `Input the comment ID` would be the Comment block, and the Comment Content block wouldn't be rendered.
2. You are editing a template / template part / etc. that would be rendered inside a Comment block. For this one, I thought you would see a generic message like `The content of the comment`.

DAreRodz replied:
> If `content` is `undefined` has two cases:
>
> 1. still loading
> 2. comment doesn't exist (this can be handled in a separate PR though and has to do with `useEntityProp`)

I see. Is there a way to know if an entity is still loading or if it doesn't exist after calling `useEntityProp`? I mean, like a value or a prop returned by that hook? That would be useful. :slightly_smiling_face:

DAreRodz replied:
> The below `if ( ! content?.rendered )` check (I had hastily added yesterday) is not correct, as it should be shown when the comment exists and there is no content.

You mean that, if `content` is not defined, it's not that the comment has no content but that it doesn't exist, right? So I guess I should write something like `if ( content && ! content.rendered )`.

ntsekouras replied:
I guess we could have a placeholder chip [like here](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/post-featured-image/edit.js#L143) when no `commentId` is provided.

>You mean that, if content is not defined, it's not that the comment has no content but that it doesn't exist,

Correct.

Probably there is a couple of things we could do for this first version, where we do not have the comment content editable - also not sure if we need to..
1. Skip the check/placeholder for a comment that exists and has no content
2. Instead of `useEntityProp`, use the `core/data` selectors like:
```
const { comment, isLoading } = useSelect(
	( select ) => {
		const { getEntityRecord, hasFinishedResolution } = select( coreStore );
		const queryArgs = [ 'root', 'comment', commentId ];
		return {
			comment: getEntityRecord( ...queryArgs ),
			isLoading: ! hasFinishedResolution( 'getEntityRecord', queryArgs ),
		};
	},
	[ commentId ]
);
```

DAreRodz committed:
Comment Content: Fix path props in icon

ntsekouras commented:
Sorry for the temp close - bad gh cli command 😄

DAreRodz commented:
Hey there, I've just created a simple schema in order to organize things better in my mind. The following diagram shows how (IMHO) the Comment Content block should behave in the context of the [Comment block](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/post-comment).

[Excalidraw link](https://excalidraw.com/#json=5677867144839168,msB3EM90hB02ijZaigqeHg)
![image](https://user-images.githubusercontent.com/6917969/136053083-cb36e70c-d8c1-4e1f-9499-5951f1bbc7fd.png)

While doing so, I came up with some questions:

1. Maybe we can use the selector you proposed (https://github.com/WordPress/gutenberg/pull/35183#discussion_r721260123) in the parent block instead, and handle there if we should render a loading/warning message when the entity is not ready or does not exist, skipping the children blocks. Would that make sense?
2. If so, should those changes be done in a different PR?
3. Am I overthinking here? Please, let me know. :sweat_smile:

Apart from that, I've also been thinking about when the Comment block is rendered inside the Comment List block. In that case, I guess the Comment block should know whether it's being rendered inside or outside a Comment List. Is there a way for doing so, or would it need to be two different blocks?

[Excalidraw link](https://excalidraw.com/#json=6451719923826688,oyevUZgka-beItzLO3L2rw)
![image](https://user-images.githubusercontent.com/6917969/136061337-3b854cfd-2fe2-49a7-bbd9-9f7d9f1a1b91.png)

DAreRodz committed:
Comment Content: Check if the comment is ready

DAreRodz committed:
Comment Content: Replace loading message with placeholder

DAreRodz commented:
I've done the changes that @ntsekouras suggested in https://github.com/WordPress/gutenberg/pull/35183#discussion_r721260123. About the placeholder chip, I guess I would need some help here from the designers (cc: @jameskoster ). :blush:

Also, if anyone can take a look at what I've posted right before, that would be great. :+1:

DAreRodz commented on packages/block-library/src/post-comment-content/edit.js:
```diff
@@ -3,29 +3,77 @@
+		return (
+			<div { ...blockProps }>
+				<Warning>
+					{ __( 'The specified comment does not exist.' ) }
```
Should I change the warning message to include the block name? Just to make it more meaningful.

```suggestion
					{ __( 'Comment Content: The specified comment does not exist.' ) }
```

jameskoster commented:
Hmm, I'm a little confused as to why we need this state:

<img width="689" alt="Screenshot 2021-10-06 at 14 51 36" src="https://user-images.githubusercontent.com/846565/136216187-cc66fcae-711b-4666-9652-816a1d90a30e.png">

The UX is particularly poor because if I choose a bad ID when inserting the Post Comment block then I need to manually remove everything to start over.

Shouldn't we run the ID validation on this save button instead?

<img width="356" alt="Screenshot 2021-10-06 at 14 53 04" src="https://user-images.githubusercontent.com/846565/136216411-cc7edd83-a95b-4d13-a011-43341e76f7f3.png">

If this "The specified comment does not exist." state is just an interim until we can do that, then I suppose it's ok. But ultimately I don't think a user should ever actually see this? 🤔

mtias commented:
A comment could be deleted, or the ID could be modified using the html editor, or coming from a different template, WordPress import, etc.

jameskoster commented:
Makes sense. In that case I think it's ok for now, but we might want to revisit holistically later.

Just to clarify @DAreRodz, this is the "placeholder chip" you were referring to, correct? :D

DAreRodz commented:
Thanks for your feedback! 🙌

@jameskoster I agree with you: the ID validation should run on the save button of the Comment block (actually, it's more or less what I posted in the previous diagram 😅). I can implement it (if that's okay with you), although I'm not sure if it's outside the scope of this PR. 🙃

> A comment could be deleted, or the ID could be modified using the html editor, or coming from a different template, WordPress import, etc.

@mtias good point, I didn't think of that. Wondering if it would make sense to show a similar message in the SSR from PHP. :thinking: 

> Just to clarify @DAreRodz, this is the "placeholder chip" you were referring to, correct? :D

Nope, I was referring to [this one](https://github.com/DAreRodz/gutenberg/blob/04abec7a4b9f8e1bd7244c378459c5bd71a7077b/packages/block-library/src/post-comment-content/edit.js#L43-L51) I've added to show something while the comment is being loaded, as suggested in https://github.com/WordPress/gutenberg/pull/35183#discussion_r721260123, but he was proposing to use it when `commentId` is undefined. I've just realized I misunderstood what Tsekouras was proposing ― sorry about that. :see_no_evil: I think he meant to use the "placeholder chip" [here](https://github.com/DAreRodz/gutenberg/blob/04abec7a4b9f8e1bd7244c378459c5bd71a7077b/packages/block-library/src/post-comment-content/edit.js#L53-L58) instead.

jameskoster commented:
> although I'm not sure if it's outside the scope of this PR.

Yeah let's do it separately.

> I think he meant to use the "placeholder chip" here instead.

Ah, the no-context placeholder :D

There's an ongoing initiative to simplify the styling of these. The Post Content block for example just [outputs a paragraph now](https://github.com/WordPress/gutenberg/pull/35243). I think we can do the same here.

DAreRodz committed:
Comment Content: Support for `textAlign` attribute

I've also removed some cases that I think
should be handled in the parent block:
- When the comment is not ready yet
- When the comment does not exist

mtias commented:
> Wondering if it would make sense to show a similar message in the SSR from PHP.

It seems for actual rendering on the front it should skip doing anything if an ID is not matched?

DAreRodz commented:
> It seems for actual rendering on the front it should skip doing anything if an ID is not matched?

It's what it does right now, I'll leave as it is then. :+1:

DAreRodz committed:
Comment Content: Add padding support

DAreRodz commented:
I've updated the PR description to reflect the changes I've done. Not sure if there's anything I missed, so asking for reviewers. :slightly_smiling_face:

SantosGuillamot commented:
I've been taking a look from the user point of view and it looks great so far 🙂 

I just have one comment. I've seen that in other PRs like the [Comment Date block](https://github.com/WordPress/gutenberg/pull/35112) and the [Comment Author block](https://github.com/WordPress/gutenberg/pull/35595), we have implemented a few more typography options, and I was wondering if we should add them here as well to keep consistency between blocks. These are the Typography options in the mentioned blocks:

- Font family
- Font size
- Appearance
- Line height
- Letter case
- Letter spacing

If I am not mistaken, right now only the Font Size, Line Height and Letter Spacing are implemented, but maybe the Font family and Appearance could make sense as well. What do you think?

Apart from that, @gziolo could you take a look at the code please?

DAreRodz commented:
> If I am not mistaken, right now only the Font Size, Line Height and Letter Spacing are implemented, but maybe the Font family and Appearance could make sense as well. What do you think?

Sure, thanks for the feedback. I'll check the PRs you mentioned and add support for those options too. :slightly_smiling_face:

DAreRodz committed:
Comment Content: Add more typography settings

Added experimental settings for changing
- Font family
- Font weight
- Font style
- Text transform
- Letter spacing

gziolo commented on packages/block-library/src/post-comment-content/edit.js:
```diff
@@ -1,32 +1,78 @@
+
+	return (
+		<>
+			{ /* Add controls to handle text alignment. */ }
```
It is implemented so the comment can be removed:
```suggestion
```

gziolo commented on packages/block-library/src/post-comment-content/edit.js:
```diff
@@ -1,32 +1,78 @@
+	attributes: { textAlign },
+	context: { commentId },
+} ) {
+	// Generate the block props adding extra classes.
```
Do we still need all of the inline code comments? Some of them document statements that at least for me can be read from the code.

gziolo commented on packages/block-library/src/post-comment-content/index.js:
```diff
@@ -1,7 +1,7 @@
  * WordPress dependencies
  */
-import { postContent as icon } from '@wordpress/icons';
+import { commentContent as icon } from '@wordpress/icons';
```
At least for now, we might want to keep the `post` prefix for consistency. It's probably better to remove this prefix in one batch if we decide to do so. I guess it would be also fine to stop referring to those blocks as Post Comment only in the UI and use the Comment form while keeping `post-comment-` internally.

gziolo reviewed (approved):
Nice work. The code looks good. I left some minor comments to consider, but they aren't blockers.

ntsekouras commented on packages/block-library/src/post-comment-content/edit.js:
```diff
@@ -1,32 +1,78 @@
+ * @param {Object} props.context              Inherited context.
+ * @param {string} props.context.commentId    The comment ID.
+ *
+ * @return {JSX.Element} React element.
```
Do we need these? 🤔  We also have this documentation about [edit/save](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-edit-save/).

gziolo replied:
I guess it's fine to have JSDoc here. Why would it be wrong?

ntsekouras replied:
It's definitely not wrong 😄. The value here lies on reading some info about the used props - as we destructure everything in this use case? Should we have this in mind for all blocks - as there is no other block with JSDoc in `edit` function?

gziolo replied:
If we would have a base TS type for `edit` it would be definitely easier to keep this JSDoc up to date. In general, it's a lot of maintenance work to keep the JSDoc entry detailed enough, but I guess we can give it a try. The fact that other blocks don't have it, it doesn't necessarily  mean it was done on purpose 😅

ntsekouras reviewed (approved):
Looks good! Thanks for your work here @DAreRodz !

jameskoster commented:
Nice! Line-height isn't working for me though. I see it is applied in the CSS, but the paragraphs in the comment content markup have a theme editor style applied with a higher priority. I suspect other type controls may have awkward interactions with different themes. Anything we can do there?

> Create a template containing just the Comment Content block, and confirm that the fallback message (The content of the comment.) is shown.

I'm just seeing an empty block – no message. I tried both the the post editor and the site editor. In this case, let's just echo the block name, that will align with the [recent work](https://github.com/WordPress/gutenberg/pull/35517) to unify these placeholders.

DAreRodz commented:
I'll do all the changes proposed. Thanks for your reviews. :relaxed:

DAreRodz committed:
Comment Content: Remove unnecessary comments

DAreRodz commented:
> Line-height isn't working for me though. I see it is applied in the CSS, but the paragraphs in the comment content markup have a theme editor style applied with a higher priority.

I see, I'm using TT1 and it works fine with this theme. About CSS and priority, I don't know if there is a better solution than trying to make the CSS rules for type controls more restrictive ― which is far from a perfect solution as themes can use more restrictive rules as well. :shrug: I would love to read other opinions on this.

> I'm just seeing an empty block – no message. I tried both the the post editor and the site editor. In this case, let's just echo the block name, that will align with the [recent work](https://github.com/WordPress/gutenberg/pull/35517) to unify these placeholders.

Oh, I guess you tried with an invalid `commentId`, am I right? I think I haven't covered that case. :sweat_smile: The placeholder appears when `commentId` is not defined, as in the case of a template part (image below).

![image](https://user-images.githubusercontent.com/6917969/137876798-bcd3687a-5025-4860-87cd-c337f91e48c5.png)

I can show the placeholder for all cases, i.e. no `commentId` and wrong `commentId`. And yup, for consistency, I'll replace the current placeholder by the block name. That's a good idea. :smiley:

DAreRodz committed:
Icons: Add `post` prefix to Comment Content icon

DAreRodz committed:
Comment Content: Show block name as placeholder

DAreRodz commented:
I think everything is ready now. :blush:

gziolo reviewed (approved):
Code still looks good 👍🏻

DAreRodz commented:
Did I break something? There's a test failing that hadn't failed before. :eyes:

gziolo commented:
> Did I break something? There's a test failing that hadn't failed before. 👀

It might be something with the env on CI:

<img width="735" alt="Screen Shot 2021-10-19 at 12 04 32" src="https://user-images.githubusercontent.com/699132/137888796-c61e807e-0edb-408d-a2d5-8265050cc75c.png">

We need to check other PRs to confirm.

Update: It's unrelated to this PR. I see the same issue in other places.

jameskoster commented:
> I see, I'm using TT1 and it works fine with this theme. About CSS and priority, I don't know if there is a better solution than trying to make the CSS rules for type controls more restrictive ― which is far from a perfect solution as themes can use more restrictive rules as well. 🤷 I would love to read other opinions on this.

This is important to get right before merge imo, because in some rare cases _none_ of these settings would work and the block would feel broken. 

Ultimately if you're changing a setting on the block, it's because you don't like what is supplied by the theme, so I feel the block should take priority here if possible.

> Oh, I guess you tried with an invalid commentId, am I right?

Correct :D It's working now 👍

geriux commented:
> Did I break something? There's a test failing that hadn't failed before. 👀

Hey there 👋 the fix for that issue is now merged into `trunk`

gziolo commented:
> > I see, I'm using TT1 and it works fine with this theme. About CSS and priority, I don't know if there is a better solution than trying to make the CSS rules for type controls more restrictive ― which is far from a perfect solution as themes can use more restrictive rules as well. 🤷 I would love to read other opinions on this.
> 
> This is important to get right before merge imo, because in some rare cases _none_ of these settings would work and the block would feel broken.
> 
> Ultimately if you're changing a setting on the block, it's because you don't like what is supplied by the theme, so I feel the block should take priority here if possible.

I’m still unsure if that is a fixable issue with code. Block themes should leave the styling to blocks when it is supported. Maybe it’s only specific to a theme you tested with and we can sort it out by publishing a dev note for theme authors. They might be able to use `theme.json` to disable the UI controls if they have opinionated rules or remove those conflicting rules from CSS. It looks like the chicken or the egg dilemma 😅

gziolo commented:
I found an edge case. I provided a comment id that doesn't exist. I was able to style everything nicely in the editor but it produces a broken HTML on the frontend:


https://user-images.githubusercontent.com/699132/138081009-c21ff6e6-882a-4b64-8082-c8815b09d5dd.mov

Should it print nothing on the frontend instead?

gziolo commented:
It tests well when using TT1 blocks theme:

Editor:

<img width="685" alt="Screen Shot 2021-10-20 at 13 05 12" src="https://user-images.githubusercontent.com/699132/138081810-0461c4e2-b283-410a-9ead-9ef3bd0f62cf.png">

Frontend:

<img width="662" alt="Screen Shot 2021-10-20 at 13 05 21" src="https://user-images.githubusercontent.com/699132/138081834-bd8aed18-94a2-4ae0-8aa3-82086ee38b9c.png">

jameskoster commented:
> Maybe it’s only specific to a theme you tested with and we can sort it out by publishing a dev note for theme authors.

That seems fair.
